Add X11DisplayManager to get display information

jkim-julie opened the pull request:
X11NativeDisplayDelegate owns X11DisplayManager and it's
registered as an X11DisplayManager's observer and it starts to
communicate with it. When X11DisplayManager is created, it
gathers display information and create DisplaySnapshot.
DesktopScreenOzone gets updated display through
DesktopScreenOzone::OnConfigurationChanged,
X11NativeDisplayDelegate::GetDisplays and
DesktopScreenOzone::OnHostDisplaysReady.

jkim-julie committed:
Add X11DisplayManager to get display information

X11NativeDisplayDelegate owns X11DisplayManager and it's
registered as an X11DisplayManager's observer and it starts to
communicate with it. When X11DisplayManager is created, it
gathers display information and create DisplaySnapshot.
DesktopScreenOzone gets updated display through
DesktopScreenOzone::OnConfigurationChanged,
X11NativeDisplayDelegate::GetDisplays and
DesktopScreenOzone::OnHostDisplaysReady.

jkim-julie commented:
@msisov, I brought the code for getting display information from DesktopScreenX11, moved DisplayShapshot to X11DisplayManager from X11NativeDisplayDelegate and added X11DisplayManager::SetDisplaySnapshot(). PTAL.

msisov commented on ui/ozone/platform/x11/BUILD.gn:
```diff
@@ -32,6 +32,8 @@ source_set("x11") {
     "x11_window_ozone.h",
     "x11_native_display_delegate.h",
     "x11_native_display_delegate.cc",
+    "x11_display_manager.cc",
```
x11_display_manager_ozone

msisov commented on ui/ozone/platform/x11/x11_display_manager.cc:
```diff
@@ -0,0 +1,335 @@
+constexpr int default_refresh = 60;
+
+// static
+gfx::ICCProfile GetICCProfileForMonitor(int monitor) {
```
Can we share this and some code below with the stock x11? Just copying the code doesn't look nice. Otherwise, you can do it while upstreaming the patch.  up to you

msisov commented on ui/ozone/platform/x11/x11_display_manager.cc:
```diff
@@ -0,0 +1,335 @@
+      xrandr_event_base_(0),
+      weak_factory_(this) {
+  // This shouldn't be called twice.
+  DCHECK(!current_snapshot_);
```
ctor won't be called twice :)))

msisov commented on ui/ozone/platform/x11/x11_display_manager.cc:
```diff
@@ -0,0 +1,335 @@
+}
+
+bool X11DisplayManager::CanDispatchEvent(const ui::PlatformEvent& event) {
+  // TODO(msisov, jkim): implement it
```
s/implement it/ implement this

msisov commented on ui/ozone/platform/x11/x11_display_manager.cc:
```diff
@@ -0,0 +1,335 @@
+}
+
+uint32_t X11DisplayManager::DispatchEvent(const ui::PlatformEvent& event) {
+  // TODO(msisov, jkim): implement it
```
s/implement it/ implement this

msisov replied:
and don't forget to add a bug address when upstreaming.

msisov commented on ui/ozone/platform/x11/x11_display_manager.cc:
```diff
@@ -0,0 +1,335 @@
+
+std::vector<display::Display> X11DisplayManager::BuildDisplaysFromXRandRInfo() {
+  DCHECK(xrandr_version_ >= 103);
+  std::vector<display::Display> displays;
```
why can't you create display snapshots here directly and avoid creating display::Display?

jkim-julie commented:
@msisov, I updated the patch with addressing your comment.
- moves 'GetICCProfileForMonitor' to x11_util.cc to share it with desktop X11.
- keeps DisplaySnapshot and DisplayMode rather than Display to avoid creating display::Display and removes unnecessary information for DisplaySnapshot.
- introduces new types to keep DisplaySnapshot and DisplayMode and uses display_id as a key value to find data.
- adds TODO to GetDeviceScaleFactor() because we have to consider device scale factors. On Desktop X11, it gets from GtkUi::GetRawDeviceScaleFactor().
- modified TODO comment.
PTAL.

msisov commented on ui/base/x/x11_util.h:
```diff
@@ -297,6 +298,7 @@ UI_BASE_X_EXPORT bool IsX11WindowFullScreen(XID window);
 // Returns true if the window manager supports the given hint.
 UI_BASE_X_EXPORT bool WmSupportsHint(XAtom atom);
 
+UI_BASE_X_EXPORT gfx::ICCProfile GetICCProfileForMonitor(int monitor);
```
add a comment.
add a space after this definition.

msisov commented on ui/ozone/platform/x11/x11_display_manager_ozone.cc:
```diff
@@ -0,0 +1,266 @@
+
+double GetDeviceScaleFactor() {
+  float device_scale_factor = 1.0f;
+  // TODO(msisov, jkim): Consider the device scale factor.
```
Can you explain what else is needed here?

msisov commented on ui/ozone/platform/x11/x11_display_manager_ozone.cc:
```diff
@@ -0,0 +1,266 @@
+  return device_scale_factor;
+}
+
+gfx::Rect GetBounds(float device_scale_factor,
```
Does it require to get bounds in dips?

msisov reviewed (approved):
I guess this is ready, isn't it?

jkim-julie commented:
Yes. right.
